File: src/com/medium/LetterCombinationsOfAPhoneNumber_17.java
/**
 * 
 */
package com.medium;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class LetterCombinationsOfAPhoneNumber_17 {
    // 2019年8月7日
    // Use DFS
    // Time Complexity: O(4^n)
    // Space Complexity:O(4^n + n)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 36.2 MB, less than 98.63%
    private static final String[] KEYS = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };

    public List<String> letterCombinations(String digits) {
        List<String> mResult = new ArrayList<String>();
        if (digits == null || digits.length() == 0) {
            return mResult;
        }
        StringBuilder mBuilder = new StringBuilder();
        DFS(digits, 0, mBuilder, mResult);

        return mResult;
    }

    private void DFS(String digits, int mOffset, StringBuilder mBuilder, List<String> mResult) {
        if (mOffset == digits.length()) {
            mResult.add(mBuilder.toString());
            return;
        }

        String mLetters = KEYS[digits.charAt(mOffset) - '0'];
        for (int i = 0; i < mLetters.length(); i++) {
            int mLength = mBuilder.length();
            DFS(digits, mOffset + 1, mBuilder.append(mLetters.charAt(i)), mResult);
            mBuilder.setLength(mLength);
        }
    }

    // 2019年12月2日
    // Use DFS 2nd
    // Time Complexity: O(4^n)
    // Space Complexity:O(4^n + n)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 35.9 MB, less than 98.63%
    public List<String> letterCombinations4(String digits) {
        List<String> mResult = new ArrayList<>();
        if(digits == null || digits.length() == 0) {
            return mResult;
        }
        DFS(digits, 0, "", mResult);
        
        return mResult;
    }
    
    private void DFS(String digits,int mOffset, String tmp, List<String> mResult) {
        if(mOffset == digits.length()) {
            mResult.add(tmp);
            return;
        }
        String mLetters = KEYS[digits.charAt(mOffset) - '0'];
        for(char c: mLetters.toCharArray()) {
            DFS(digits, mOffset + 1, tmp + c, mResult);
        }
    }

    // 2019年8月7日
    // Use BFS
    // Time Complexity: O(4^n)
    // Space Complexity:O(2*(4^n))
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 36.2 MB, less than 98.63%
    public List<String> letterCombinations2(String digits) {
        LinkedList<String> mResult = new LinkedList<>();
        if (digits.isEmpty()) {
            return mResult;
        }
        mResult.add(""); // avoid null pointer
        while (mResult.peek().length() != digits.length()) {
            String mRemove = mResult.remove();
            String mMap = KEYS[digits.charAt(mRemove.length()) - '0'];
            for (char c : mMap.toCharArray()) {
                mResult.addLast(mRemove + c);
            }
        }
        return mResult;
    }

    // 2019年12月2日
    // Use BFS 2nd
    // Time Complexity: O(4^n)
    // Space Complexity:O(2*(4^n))
    // Runtime: 1 ms, faster than 59.45%
    // Memory Usage: 35.7 MB, less than 98.63%
    public List<String> letterCombinations3(String digits) {
        LinkedList<String> mResult = new LinkedList<>();
        if (digits.isEmpty()) {
            return mResult;
        }
        mResult.add(""); // avoid null pointer

        for (char digit : digits.toCharArray()) {
            LinkedList<String> mTmp = new LinkedList<>();
            for (String mStr : mResult) {
                for (char c : KEYS[digit - '0'].toCharArray()) {
                    mTmp.add(mStr + c);
                }
            }
            mResult = (LinkedList<String>) mTmp.clone();
            mTmp.clear();
        }
        return mResult;
    }
}

File: Cpp_src/medium/letter_combinations_of_a_phone_number_17.cc
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

// Use DFS
// Time Complexity: O(4^n)
// Space Complexity:O(4^n + n)
// Runtime: 4 ms, faster than 60.35%
// Memory Usage: 8.5 MB, less than 88.57%
class Solution
{
public:
    vector<string> letterCombinations(string digits)
    {
        vector<string> result;
        if (digits.size() == 0)
        {
            return result;
        }

        dfs(digits, 0, "", result);
        return result;
    }

private:
    string KEYS[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    void dfs(string digits, int offset, string tmp, vector<string> &result)
    {
        if (offset == digits.size())
        {
            result.push_back(tmp);
            return;
        }

        string letters = KEYS[digits[offset] - '0'];
        for (int i = 0; i < letters.size(); i++)
        {
            dfs(digits, offset + 1, tmp + letters[i], result);
        }
    }
};

// Use BFS
// Time Complexity: O(4^n)
// Space Complexity:O(2*(4^n))
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 8.4 MB, less than 100.00%
class Solution2
{
public:
    vector<string> letterCombinations(string digits)
    {
        vector<string> result;
        if (digits.empty())
        {
            return result;
        }
        result.push_back("");
        for (char digit : digits)
        {
            vector<string> tmp;
            for (string s : result)
            {
                for (char c : KEYS[digit - '0'])
                {
                    tmp.push_back(s + c);
                }
            }
            result.swap(tmp);
        }
        return result;
    }

private:
    string KEYS[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
};